File: android/app/src/main/java/com/aweffr/kotlinconf/shared/RNShared.kt
package com.aweffr.kotlinconf.shared

import android.util.Log
import com.aweffr.shared.Todo
import com.aweffr.shared.TodoService
import com.aweffr.shared.TodoServiceEventPayload
import com.facebook.react.ReactPackage
import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.NativeModule
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.WritableMap
import com.facebook.react.modules.core.DeviceEventManagerModule
import com.facebook.react.uimanager.ViewManager
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.launch

private const val TAG = "RNShared"

private fun debugLog(message: String) {
    val threadId = Thread.currentThread().id
    val threadName = Thread.currentThread().name
    Log.i(TAG, "[threadId:$threadId, threadName:$threadName]$message")
}

fun Todo.asReactMap(): WritableMap {
    return Arguments.createMap().apply {
        putInt("id", id)
        putString("title", title)
        putBoolean("completed", completed)
    }
}

class RNSharedPackage : ReactPackage {
    override fun createNativeModules(reactContext: ReactApplicationContext): List<NativeModule> {
        return listOf(RNSharedModule(reactContext))
    }

    override fun createViewManagers(reactContext: ReactApplicationContext): List<ViewManager<*, *>> {
        return emptyList()
    }
}

class RNSharedModule(reactContext: ReactApplicationContext) : ReactContextBaseJavaModule(reactContext) {
    override fun getName(): String {
        return TAG
    }

    private val todoService = TodoService()
    private val scope: CoroutineScope
        get() = todoService.scope


    @ReactMethod
    fun getLocalTodos(promise: Promise) {
        debugLog("getLocalTodos called.")
        try {
            val todos = todoService.getLocalTodos()
            val result = Arguments.createArray().apply {
                todos.forEach {
                    pushMap(it.asReactMap())
                }
            }
            promise.resolve(result)
        } catch (e: Exception) {
            promise.reject("RNSharedError", e)
        }
    }

    @ReactMethod
    fun getHttpTodos(promise: Promise) {
        debugLog("getHttpTodos called.")
        scope.launch {
            try {
                val todos = todoService.getHttpTodos()
                val result = Arguments.createArray().apply {
                    todos.forEach {
                        pushMap(it.asReactMap())
                    }
                }
                promise.resolve(result)
            } catch (e: Exception) {
                promise.reject("RNSharedError", e)
            }
        }
        debugLog("getHttpTodos finished.")
    }

    @ReactMethod
    fun getHttpTodosAndSendEvent(taskId: String) {
        todoService.getHttpTodosAndSendEvent(taskId)
    }

    @ReactMethod
    fun addListener(eventName: String) {
    }

    @ReactMethod
    fun removeListeners(count: Int) {
    }

    private val todosReceivedCallback: (TodoServiceEventPayload) -> Unit = { payload ->
        debugLog("todosReceivedCallback called. payload: $payload")
        val jsPayload = Arguments.createMap().apply {
            putString("taskId", payload.taskId)
            putString("message", payload.message)
            putArray("todos", Arguments.createArray().apply {
                payload.todos.forEach {
                    pushMap(it.asReactMap())
                }
            })
        }
        reactApplicationContext
            .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
            .emit(TodoService.EVENT_NAME_TODOS_RECEIVED, jsPayload)
    }

    init {
        todoService.registerSendEventCallback(todosReceivedCallback)
    }

    override fun getConstants(): MutableMap<String, Any>? {
        return mutableMapOf(
            "EVENT_NAME_TODOS_RECEIVED" to TodoService.EVENT_NAME_TODOS_RECEIVED
        )
    }
}

File: ios/RNShared.swift
//
//  RNShared.swift
//  ReactNativeWithKMP
//
//  Created by MacPro2 on 2024/6/27.
//

import Foundation
import shared

@objc(RNShared)
class RNShared: RCTEventEmitter {

  override static func requiresMainQueueSetup() -> Bool {
    return true
  }

  var todoService = TodoService()

  private func todoAsReactMap(todo: Todo) -> [String: Any] {
    return ["id": todo.id, "title": todo.title, "completed": todo.completed]
  }


  @objc(getLocalTodos:rejecter:)
  func getLocalTodos(_ resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
    let todos = todoService.getLocalTodos()
    let jsPayload = todos.map { todo in
      todoAsReactMap(todo: todo)
    }
    resolve(jsPayload)
  }

  @objc(getHttpTodos:rejecter:)
  func getHttpTodos(_ resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
    DispatchQueue.main.async {
      print("getTodosHttp called! current thread: \(Thread.current)")
      self.todoService.getHttpTodos { todos, error in
        print("getTodosHttp finished! current thread: \(Thread.current)")
        if let err = error {
          reject("RNSharedError", "getHttpTodos fail", err)
        } else if let _todos = todos {
          let result = _todos.map { todo in
            ["id": todo.id, "title": todo.title, "completed": todo.completed]
          }
          print("Got todos count: \(result.count)")
          resolve(result)
        } else {
          reject("RNSharedError", "getHttpTodos fail. todos is nil", nil)
        }
      }
    }
  }
  
  @objc(getHttpTodosAndSendEvent:)
  func getHttpTodosAndSendEvent(_ taskId: String) {
    todoService.getHttpTodosAndSendEvent(taskId: taskId)
  }

  override init() {
    super.init()
    let greetStr = Greeting().greet()
    print("=====\nRNShared init called! greetStr:\(greetStr)\n=====\n")

    todoService.registerSendEventCallback { payload in
      print("callback fired from shared module! \(Thread.current)")
      let jsPayload = [
        "taskId": payload.taskId,
        "message": payload.message,
        "todos": payload.todos.map { todo in
          self.todoAsReactMap(todo: todo)
        }
      ]
      self.sendEvent(withName: TodoService.companion.EVENT_NAME_TODOS_RECEIVED, body: jsPayload)
    }
  }
  
  override func supportedEvents() -> [String]! {
    return [TodoService.companion.EVENT_NAME_TODOS_RECEIVED]
  }

  override func constantsToExport() -> [AnyHashable : Any]! {
    return ["EVENT_NAME_TODOS_RECEIVED": TodoService.companion.EVENT_NAME_TODOS_RECEIVED]
  }
}
